crates/biome_js_analyze/src/lint/nursery/no_substr.rs

```diff
@@ -0,0 +1,132 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, ActionCategory, Ast, FixKind, Rule, RuleDiagnostic,
+};
+use biome_console::markup;
+use biome_js_factory::make;
+use biome_js_syntax::JsCallExpression;
+use biome_rowan::{AstSeparatedList, BatchMutationExt, TextRange, TokenText};
+
+use crate::JsRuleAction;
+
+declare_rule! {
+    /// Enforce the use of `slice()` over `substr()` and `substring()`.
+    ///
+    /// `slice()` is preferred over `substr()` and `substring()` because it is a more popular option with clearer behavior,
+    ///  and it has a consistent counterpart in arrays.
+    ///
+    /// Note that `substr`, `substring` and `slice` are not identical when arguments are passed.
+    /// For detailed differences, refer to the MDN documentation:
+    /// - [The difference between substring() and substr()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#the_difference_between_substring_and_substr)
+    /// - [Differences between substring() and slice()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#differences_between_substring_and_slice)
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substr(start, length);
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo.substring(indexStart, indexEnd);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// foo.slice(beginIndex, endIndex);
+    /// ```
+    ///
+    pub NoSubstr {
+        version: "1.8.0",
+        name: "noSubstr",
+        language: "js",
+        recommended: false,
+        fix_kind: FixKind::Safe,
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct NoSubstrState {
+    member_name: TokenText,
+    span: TextRange,
+    replaced_member_name: &'static str,
+    has_arguments: bool,
+}
+
+impl Rule for NoSubstr {
+    type Query = Ast<JsCallExpression>;
+    type State = NoSubstrState;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let node = ctx.query();
+        let callee = node.callee().ok()?;
+        let expression = callee.as_js_static_member_expression()?;
+        let value_token = expression.member().ok()?.value_token().ok()?;
+        let string_function_name = value_token.text_trimmed();
+        let arguments = node.arguments().ok()?;
+        let args = arguments.args();
+
+        if matches!(string_function_name, "substr" | "substring") {
+            Some(NoSubstrState {
+                member_name: value_token.token_text_trimmed(),
+                span: value_token.text_range(),
```

Comment:
we can simplify things a bit by just passing value_token in state. You will no longer need to unwrap them multiple times in action I think you can directly replace `value_token` with `make::indent` and in diagnostics call text_range there.